src/bun.js/readline.exports.js

```diff
@@ -0,0 +1,3122 @@
+  StringIteratorPrototypeNext,
+);
+
+// ----------------------------------------------------------------------------
+// Section: "Internal" modules
+// ----------------------------------------------------------------------------
+
+/**
+ * Returns true if the character represented by a given
+ * Unicode code point is full-width. Otherwise returns false.
+ */
+var isFullWidthCodePoint = (code) => {
+  // Code points are partially derived from:
+  // https://www.unicode.org/Public/UNIDATA/EastAsianWidth.txt
+  return (
+    code >= 0x1100 &&
+    (code <= 0x115f || // Hangul Jamo
+      code === 0x2329 || // LEFT-POINTING ANGLE BRACKET
+      code === 0x232a || // RIGHT-POINTING ANGLE BRACKET
+      // CJK Radicals Supplement .. Enclosed CJK Letters and Months
+      (code >= 0x2e80 && code <= 0x3247 && code !== 0x303f) ||
+      // Enclosed CJK Letters and Months .. CJK Unified Ideographs Extension A
+      (code >= 0x3250 && code <= 0x4dbf) ||
+      // CJK Unified Ideographs .. Yi Radicals
+      (code >= 0x4e00 && code <= 0xa4c6) ||
+      // Hangul Jamo Extended-A
+      (code >= 0xa960 && code <= 0xa97c) ||
+      // Hangul Syllables
+      (code >= 0xac00 && code <= 0xd7a3) ||
+      // CJK Compatibility Ideographs
+      (code >= 0xf900 && code <= 0xfaff) ||
+      // Vertical Forms
+      (code >= 0xfe10 && code <= 0xfe19) ||
+      // CJK Compatibility Forms .. Small Form Variants
+      (code >= 0xfe30 && code <= 0xfe6b) ||
+      // Halfwidth and Fullwidth Forms
+      (code >= 0xff01 && code <= 0xff60) ||
+      (code >= 0xffe0 && code <= 0xffe6) ||
+      // Kana Supplement
+      (code >= 0x1b000 && code <= 0x1b001) ||
+      // Enclosed Ideographic Supplement
+      (code >= 0x1f200 && code <= 0x1f251) ||
+      // Miscellaneous Symbols and Pictographs 0x1f300 - 0x1f5ff
+      // Emoticons 0x1f600 - 0x1f64f
+      (code >= 0x1f300 && code <= 0x1f64f) ||
+      // CJK Unified Ideographs Extension B .. Tertiary Ideographic Plane
+      (code >= 0x20000 && code <= 0x3fffd))
+  );
+};
+
+var isZeroWidthCodePoint = (code) => {
+  return (
+    code <= 0x1f || // C0 control codes
+    (code >= 0x7f && code <= 0x9f) || // C1 control codes
+    (code >= 0x300 && code <= 0x36f) || // Combining Diacritical Marks
+    (code >= 0x200b && code <= 0x200f) || // Modifying Invisible Characters
+    // Combining Diacritical Marks for Symbols
+    (code >= 0x20d0 && code <= 0x20ff) ||
+    (code >= 0xfe00 && code <= 0xfe0f) || // Variation Selectors
+    (code >= 0xfe20 && code <= 0xfe2f) || // Combining Half Marks
+    (code >= 0xe0100 && code <= 0xe01ef)
+  ); // Variation Selectors
+};
+
+/**
+ * Returns the number of columns required to display the given string.
+ */
+var getStringWidth = function getStringWidth(str, removeControlChars = true) {
+  let width = 0;
+
+  if (removeControlChars) str = stripVTControlCharacters(str);
+  str = StringPrototypeNormalize.call(str, "NFC");
+  for (const char of new SafeStringIterator(str)) {
```

Comment:
Can this use `Buffer.byteLength`?

Reply:
I tried this but the tests failed. I realized it's not really the byte length like it seems but the column width in the terminal which varies among characters-- some utf-8 characters are two columns wide even though they are represented by 1 byte. 

I guess you have to manually look up the width based on the char code... It seems like they mostly correspond to ranges. Could be some bitwise hack to calculate that for the whole string at once, but that may be an overkill of an optimization.

Reply:
https://stackoverflow.com/a/9145712
This seems to have the general idea of how the calculation works across the whole spec, but I think it's been updated a few times since this original spec.

Reply:
oh interesting, didn't know about this